EZP-27508: As a Developer I want an API that removes a translation from a version

alongosz opened the pull request:
# Implements [EZP-27508](https://jira.ez.no/browse/EZP-27508)

This PR implements PHP API method:
```
ContentService::deleteTranslationFromDraft(VersionInfo $versionInfo, $languageCode);
```
 which deletes the Translation specified by a language code from a Content Object Version Draft.

Related PRs:
- #2001 - initial discussions for similar feature removing Translation from the entire Content, naming introduced there will be changed according to [this conclusion](https://github.com/ezsystems/ezpublish-kernel/pull/2081#issuecomment-334177307)
- #2086 - REST endpoint using PHP API implemented by this PR.

Note: It's a responsibility of PHP API Consumer to first create Draft and then publish it after removing language. Moreover, if a Translation is the main one, it needs to be changed manually using `ContentService::updateContentMetadata` first.

**TODO**:
- [x] Create spec for `ContentService::deleteTranslation` method.
- [x] See if what spec describes is technically possible (issue with merging field/translation when publishing)
- [x] Implement the method on PHP API interface and implementing layers.
- [x] Improve Spec.
- [x] Create integration tests.
- [x] Delete external storage data for removed fields.
- [x] Rebase and reorder commits before merge.
- [x] Fix naming of API method.
- [x] **Rebase before merge**

*Note: created branch on `ezsystems` so dependent features can temporarily target to this branch (currently it's impossible to change base fork, only base branch can be changed)*

andrerom commented on doc/specifications/translations/removal.md:
````diff
@@ -57,3 +57,35 @@ class RemoveTranslationSignal extends Signal
+```
+
+Since the returned Content Draft is to be published, both Search and HttpCache are already handled
+by `PublishVersion` Slots.
````
```
by `PublishVersion` Slots once call to publishVersion() is made.
```

alongosz replied:
Fixed in f2bf66b

alongosz commented:
Added one more fixup to Spec in e6c34a0.
As discussed, `mainLanguage` still needs to be changed manually before `deleteTranslation`, as it was in case of `removeTranslation`.
The reason for that is this property belongs to metadata of a Content but `deleteTranslation` works on Drafts (metadata is not a part of a Draft). Hence trying to handle it by `deleteTranslation` would result in inconsistent state of a Content Object.

REST API endpoint will handle updating `mainLanguage` prior to calling `deleteTranslation`.

andrerom commented:
> REST API endpoint will handle updating mainLanguage prior to calling deleteTranslation.

ok, just note that then the endpoint should take argument for what to change it to I assume, otherwise it will be random afaik.

alongosz commented:
> > REST API endpoint will handle updating mainLanguage prior to calling deleteTranslation.

> ok, just note that then the endpoint should take argument for what to change it to I assume, otherwise it will be random afaik. And if not set also return error like here.

Could be, but there is already [a suggestion on how to do it](https://jira.ez.no/browse/EZP-25684?focusedCommentId=215847&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-215847):
> I think we should just define a rule (and document it for the API), what makes sense to me is the chronological order. Take the oldest language introduced in the content item.

alongosz commented:
> On a documentation side, I'm not sure about the phpdoc as a structure in a markdown file. Is it the most usable ? After all, it is meant to be generated to documentation, not used as such in text :)

It's there because it seems to me like a good way to document thrown exceptions and expected parameter types. However the leading PhpDoc line with verbose description and also `@since` tag is really not needed there. I can remove it. Would it be a good compromise @bdunogier? Or maybe you prefer a different way to describe what I've included in PhpDoc?

> Also, would usage examples for the main use-cases make sense in such files ?

I could add them, no problem :)

// cc @andrerom

alongosz commented:
In 0b4e102 I've improved Spec [removal.md](https://github.com/ezsystems/ezpublish-kernel/blob/0b4e10291840583ec11d078362a8da29c6eab6c4/doc/specifications/translations/removal.md):
- added extra note on main Translation,
- removed unnecessary information from PhpDoc (they're in API interface anyway), however keeping parameter types and thrown exceptions. Is it ok @bdunogier?
- added example usage of `deleteTranslation`.

alongosz commented on eZ/Publish/Core/Repository/ContentService.php:
```diff
@@ -1964,6 +1964,86 @@ function (Location $location) {
+                '$languageCode',
+                'Specified Translation is the only one Content Object Version has'
+            );
+        }
```
Added also this check, as Gateway will throw exception as a last resort, but messing with state of the doctrine transaction (this is fine as long as it won't happen, so preventing it here). 
Reviewers: could be discussed if you want more details

alongosz commented on eZ/Publish/Core/Repository/ContentService.php:
```diff
@@ -1964,6 +1964,86 @@ function (Location $location) {
+            // avoid wrapping expected NotFoundException in BadStateException handled below
+            $this->repository->rollback();
+            throw $e;
+        } catch (Exception $e) {
```
It's possible that Content Handler will throw NotFoundException, so I'm avoiding wrapping this in BadStateException below.

andrerom commented on doc/specifications/translations/removal.md:
```diff
@@ -57,3 +57,53 @@ class RemoveTranslationSignal extends Signal
+translation removal.
+
+**Note**: If a removed Translation is the main one, it needs to be updated manually using
+`ContentService::updateContentMetadata` method.
```
maybe:

> **Note**: To removed main Translation, main language needs to be changed manually using
 +`ContentService::updateContentMetadata` method first.

alongosz replied:
Ok, sounds better, new spec in 714469a :)

andrerom commented on eZ/Publish/Core/Persistence/Cache/ContentHandler.php:
```diff
@@ -317,4 +317,27 @@ public function removeTranslationFromContent($contentId, $languageCode)
+        $this->cache->clear('content', $contentId, $versionNo);
+        $this->cache
+            ->getItem('content', $contentId, $versionNo, self::ALL_TRANSLATIONS_KEY)
+            ->set($content)
```
1. Actually, could you skip the cache warming?
2. versionNo is wrong here and above, cause the returned item is a new version right? and as it's not published afaik we don't need to clear any cache _(same case as `createDraftFromVersion`)_

alongosz replied:
> Also versionNo is wrong here and above right? Cause the returned content is a new draft version, and as it's not published so afaik we don't need to clear any cache (same case as createDraftFromVersion)

Version passed to this method is a draft, so `versionNo` is correct. It works in a similar way like `updateContent` from which I copied cache clearing and warming. 

However I don't get cached value while "playing" with REST requests on draft (after `deleteTranslation` and before publishing), so maybe it's really not needed there.

andrerom replied:
On point 2: My bad, I was miss reading the concepts and thinking the method here creates a new version for you w/o the given translation.

So just skip the cache warmin then, meaning only call needed is:
`this->cache->clear('content', $contentId, $versionNo);`


As for not getting cached value, unsure what you are actually testing, but if what you are testing triggers `load()` on this handler, it is cached. If it just uses ´ loadVersionInfo´ then it's not in 1.x (it is in 2.x).

alongosz replied:
Ok, fixed: https://github.com/ezsystems/ezpublish-kernel/pull/2081/commits/561938de942cf23c2549cd31e2ebd257194b452e#diff-fb693bf163e578a01367f4466a44ad9b

alongosz commented:
Fixed according to review and already squashed

Nattfarinn commented on eZ/Publish/SPI/Persistence/Content/Handler.php:
```diff
@@ -266,4 +266,15 @@ public function publish($contentId, $versionNo, MetadataUpdateStruct $metaDataUp
+     *
+     * @return \eZ\Publish\SPI\Persistence\Content The Content Draft w/o removed Translation
+     */
+    public function deleteTranslationFromVersion($contentId, $versionNo, $languageCode);
```
`public function removeTranslationFromContent($contentId, $languageCode);`
but
`public function deleteTranslationFromVersion($contentId, $versionNo, $languageCode);`

Wouldn't it be better to keep naming convention and use `remove` here as well?

Nattfarinn commented on eZ/Publish/API/Repository/ContentService.php:
```diff
@@ -427,6 +427,26 @@ public function loadTranslationInfos(ContentInfo $contentInfo, array $filter = a
+     *
+     * @since 6.12
+     */
+    public function deleteTranslation(VersionInfo $versionInfo, $languageCode);
```
Maybe `removeTranslationFromDraft`?

Having both, `deleteTranslation` and `removeTranslation`, but operating on different data will be very confusing.

alongosz replied:
> Maybe removeTranslationFromDraft?
> Having both methods, deleteTranslation and removeTranslation, but operating on different data will be very confusing.

Could be. I chose this name to align with `deleteContent` and `deleteVersion`.
`removeTranslation` refers to removing from entire content so this is nice distinction. It had been chosen due to Legacy reasons
// cc @andrerom

alongosz commented:
We're discussing this with @lserwatka and @Nattfarinn right now.
Conclusion:
1. we want consistency on naming in current API, not Legacy API
2. name for method deleting translation from single version draft should be `deleteTranslationFromDraft`
3. name for method permanently deleting translation from Content should be `deleteTranslation` (no need to add Content suffix as we're "in" `ContentService`)

Action 2. can be easily applied to existing not merged PRs, actions 1. and 3. could be applied as follow-ups stories involving deprecation of old name and adding new one.

ping @andrerom

andrerom commented:
@alongosz 👍  But make sure to follow BC approach for methods that has been in a release already.

alongosz commented:
Refactored PAPI method name to `deleteTranslationFromDraft`. See fixups:
- 1ffd087
- 54df8d4
- c442c7b

ping @andrerom @Nattfarinn

Nattfarinn commented:
![200_s](https://user-images.githubusercontent.com/5822569/31226305-e90b41f2-a9d5-11e7-9749-5e4041ccc317.gif)

alongosz commented:
Squashed fixups, sending to QA // cc @mnocon @micszo

alongosz commented:
Thanks @mnocon 
@andrerom this is mergable now :)
